src/test/java/com/billingapi/controller/InvoiceControllerTest.java

```diff
@@ -0,0 +1,586 @@
+                .discountOnSale(10.00)
+                .userId(1)
+                .isDeleted(false)
+                .timestamp(LocalDateTime.now())
+                .description("good")
+                .quantity(2)
+                .total(new BigDecimal("98000.0"))
+                .build();
+
+        given(invoiceService.findInvoiceById(invoiceId)).willReturn(newInvoice);
+
+        // when -  action or the behaviour that we are going test
+        ResultActions response = mockMvc.perform(get("/invoice/aa", invoiceId));
+
+        // then - verify the output
+        response.andExpect(status().isBadRequest())
+                .andDo(print());
+    }
+
+
+
+
+  /*  Test to Create invoice using InvoiceRequest a request Dto which giving the request to add new Invoice
+     and return  InvoiceResponse
+  */
+       @Test
+       public void givenInvoiceObject_whenCreateInvoice_thenReturnaddInvoice() throws Exception {
+                    Invoice newInvoice = Invoice.builder()
+                   .invoiceId(1)
+                   .itemName("jeans")
+                   .price(new BigDecimal("98000.00"))
+                   .tax(10.00)
+                   .vat(15.00)
+                   .discount(10.00)
+                   .isItemOnSale(true)
+                   .discountOnSale(10.00)
+                   .userId(1)
+                   .isDeleted(false)
+                   .description("good")
+                   .quantity(2)
+                   .total(new BigDecimal(205800.00))
+                   .build();
+
+
+
+
+
+   InvoiceRequest invoiceRequest = InvoiceRequest.builder()
+                .itemName("jeans")
+                .price(new BigDecimal("98000.0"))
+                .tax(10.00)
+                .vat(15.00)
+                .discount(10.00)
+                .isItemOnSale(true)
+                .discountOnSale(10.00)
+                .quantity(2)
+                .description("good")
+                .build();
+
+           BigDecimal price = invoiceRequest.getPrice();
+           double tax = invoiceRequest.getTax();
+           double vat = invoiceRequest.getVat();
+           double discount = invoiceRequest.getDiscount();
+           double discountOnSale = invoiceRequest.getDiscountOnSale();
+           int quantity = invoiceRequest.getQuantity();
+           BigDecimal newtax = new BigDecimal(tax);
+           BigDecimal newvat = new BigDecimal(vat);
+           BigDecimal newdiscount = new BigDecimal(discount);
+           BigDecimal newdiscountonsale = new BigDecimal(discountOnSale);
+           BigDecimal newquantity = new BigDecimal(quantity);
+           BigDecimal totalDiscount = ((newtax).add(newvat).subtract(newdiscount).subtract(newdiscountonsale));
+           BigDecimal subTotal = (price.multiply(totalDiscount).divide(new BigDecimal("100.0")));
+           BigDecimal total = (price.add(subTotal));
+           BigDecimal alltotal = total.multiply(newquantity);
+           invoiceRequest.setTotal(alltotal);
```

Comment:
We don't need to write to much code here. This can be conb=verted to single line.
subtotal calculation , we need to discuss ?